### PR TITLE
cpu/uart: create common uart hw fc module

### PR DESCRIPTION
**Pull request by dylad**
### Contribution description
This PR merges the ~~STM32_PERIPH_UART_HW_FC and SAM0_PERIPH_UART_HW_FC~~ STM32, SAM0, NRF, CC25xx and CC26xx UART implementation to use the generic PERIPH_UART_HW_FC so it can be reuse by other MCU families in order to provide hardware flow control support for UART driver.

### Testing procedure
CI test should be enough I guess ?

### Issues/PRs references
Change was proposed in #13001(comment)

### CODE REVIEW

**Comment by aabadie** [comment]
Nice one!
UART implementation of nrf also provides hw fc, maybe also adapt it ?

**Comment by dylad** [comment]
> UART implementation of nrf also provides hw fc, maybe also adapt it ?

Oh, I was not aware of it. I'll do it tomorrow.

**Comment by dylad** [comment]
Update UART implementation for nrf, cc25xx and cc26xx families.

File: boards/cc1352-launchpad/include/periph_conf.h [hunk 1]
```diff
@@ -71,30 +71,30 @@ static const timer_conf_t timer_config[] = {
 * The used CC26x0 CPU only supports a single UART device, so all we need to
 * configure are the RX and TX pins.
 *
-* Optionally we can enable hardware flow control, by setting UART_HW_FLOW_CTRL
-* to 1 and defining pins for UART_CTS_PIN and UART_RTS_PIN.
+* Optionally we can enable hardware flow control, by using periph_uart_hw_fc
+* module (USEMODULE += PERIPH_UART_HW_FC) and defining pins for UART_CTS_PIN
```
Comment (aabadie) [inline boards/cc1352-launchpad/include/periph_conf.h]:
```suggestion
* module (USEMODULE += periph_uart_hw_fc) and defining cts_pin and rts_pin.
```

File: boards/cc2650stk/include/periph_conf.h [hunk 2]
```diff
@@ -75,15 +75,16 @@ static const timer_conf_t timer_config[] = {
...
-     .cts_pin = 0,      /* ignored when flow_control is 0 */
-     .flow_control = 0,
-     .intn = UART0_IRQN
- }
+    {
+        .regs = UART0,
+        .tx_pin = 29,
+        .rx_pin = 28,
+#ifdef MODULE_PERIPH_UART_HW_FC
+        .rts_pin = 0,      /* ignored when flow_control is 0 */
```
Comment (aabadie) [inline boards/cc2650stk/include/periph_conf.h]:
```suggestion
        .rts_pin = GPIO_UNDEF,
```

File: boards/cc2650stk/include/periph_conf.h [hunk 3]
```diff
@@ -75,15 +75,16 @@ static const timer_conf_t timer_config[] = {
...
+        .cts_pin = 0,      /* ignored when flow_control is 0 */
```
Comment (aabadie) [inline boards/cc2650stk/include/periph_conf.h]:
```suggestion
        .cts_pin = GPIO_UNDEF,
```

File: boards/cc1352-launchpad/include/periph_conf.h [hunk 4]
```diff
@@ -71,30 +71,30 @@ static const timer_conf_t timer_config[] = {
...
-     .flow_control = 0,
-     .intn = UART0_IRQN
- }
+    {
+        .regs = UART0,
+        .tx_pin = 13,
+        .rx_pin = 12,
+#ifdef MODULE_PERIPH_UART_HW_FC
+        .rts_pin = GPIO_UNDEF,
+        .cts_pin = GPIO_UNDEF,      /* ignored when flow_control is 0 */
```
Comment (aabadie) [inline boards/cc1352-launchpad/include/periph_conf.h]:
```suggestion
        .cts_pin = GPIO_UNDEF,
```

File: cpu/nrf5x_common/periph/uart.c [hunk 5]
```diff
@@ -112,34 +113,33 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
     /* configure TX pin */
     gpio_init(UART_PIN_TX, GPIO_OUT);
     PSEL_TXD = UART_PIN_TX;
-
+    NRF_UART0->PSELRTS = 0xffffffff;        /* pin disconnected */
```
Comment (aabadie) [inline cpu/nrf5x_common/periph/uart.c]: nrf52840 variant can have 2 UARTs so I think you should keep the specific else block (nrf52840 != other nrf5x) to set RTS/CTS pin disconnected.

**Comment by dylad** [comment]
> I know the use of macro makes the code not very readable there but it was the price to pay to save a few bytes in ROM.

Maybe this situation could be improve by using runtime check with IS_USED stuff ?

**Comment by dylad** [comment]
Force-pushed to rebase on current master to get rid of some merge conflicts.

File: boards/cc1352-launchpad/include/periph_conf.h (hunk 4 above)
Comment (dylad) [reply boards/cc1352-launchpad/include/periph_conf.h]: fixed.

File: boards/cc2650stk/include/periph_conf.h (hunk 3 above)
Comment (dylad) [reply boards/cc2650stk/include/periph_conf.h]: fixed.

File: boards/cc2650stk/include/periph_conf.h (hunk 2 above)
Comment (dylad) [reply boards/cc2650stk/include/periph_conf.h]: fixed.

File: boards/cc1352-launchpad/include/periph_conf.h (hunk 1 above)
Comment (dylad) [reply boards/cc1352-launchpad/include/periph_conf.h]: fixed.

File: cpu/nrf5x_common/periph/uart.c (hunk 5 above)
Comment (dylad) [reply cpu/nrf5x_common/periph/uart.c]: I replaced `NRF_UART0` by `dev(uart)`. This way the RTS and CTS will be first disabled and only enabled if needed.

Comment (dylad) [reply cpu/nrf5x_common/periph/uart.c]: Ok I dropped those lines :)

File: boards/cc2650stk/include/periph_conf.h [hunk 6]
```diff
@@ -75,15 +75,16 @@ static const timer_conf_t timer_config[] = {
...
-     .cts_pin = 0,      /* ignored when flow_control is 0 */
-     .flow_control = 0,
-     .intn = UART0_IRQN
- }
+    {
+        .regs = UART0,
+        .tx_pin = 29,
+        .rx_pin = 28,
+#ifdef MODULE_PERIPH_UART_HW_FC
+        .rts_pin = GPIO_UNDEF,      /* ignored when flow_control is 0 */
```
Comment (aabadie) [inline boards/cc2650stk/include/periph_conf.h]:
```suggestion
        .rts_pin = GPIO_UNDEF,
```

File: boards/cc2650stk/include/periph_conf.h [hunk 7]
```diff
@@ -75,15 +75,16 @@ static const timer_conf_t timer_config[] = {
...
+        .cts_pin = GPIO_UNDEF,      /* ignored when flow_control is 0 */
```
Comment (aabadie) [inline boards/cc2650stk/include/periph_conf.h]:
```suggestion
        .cts_pin = GPIO_UNDEF,
```

File: cpu/nrf5x_common/periph/uart.c [hunk 8]
```diff
@@ -115,31 +116,29 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
 
 #ifdef CPU_MODEL_NRF52840XXAA
     /* enable HW-flow control if defined */
-    if (UART_HWFLOWCTRL) {
+ #ifdef MODULE_PERIPH_UART_HW_FC
         /* set pin mode for RTS and CTS pins */
+        if (UART_PIN_RTS != GPIO_UNDEF || UART_PIN_CTS != GPIO_UNDEF) {
```
Comment (aabadie) [inline cpu/nrf5x_common/periph/uart.c]:
```suggestion
        if (UART_PIN_RTS != GPIO_UNDEF && UART_PIN_CTS != GPIO_UNDEF) {
```

File: cpu/nrf5x_common/periph/uart.c [hunk 9]
```diff
@@ -115,31 +116,29 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+            /* configure RTS and CTS pins to use */
+            dev(uart)->PSEL.RTS = UART_PIN_RTS;
+            dev(uart)->PSEL.CTS = UART_PIN_CTS;
+            REG_CONFIG |= UART_CONFIG_HWFC_Msk; /* enable HW flow control */
+        }
+#endif
+#else
+#ifdef MODULE_PERIPH_UART_HW_FC
+    /* set pin mode for RTS and CTS pins */
+    if (UART_PIN_RTS != GPIO_UNDEF || UART_PIN_CTS != GPIO_UNDEF) {
```
Comment (aabadie) [inline cpu/nrf5x_common/periph/uart.c]:
```suggestion
    if (UART_PIN_RTS != GPIO_UNDEF && UART_PIN_CTS != GPIO_UNDEF) {
```

File: cpu/nrf5x_common/periph/uart.c [hunk 10]
```diff
@@ -115,31 +116,29 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
         gpio_init(UART_PIN_RTS, GPIO_OUT);
         gpio_init(UART_PIN_CTS, GPIO_IN);
         /* configure RTS and CTS pins to use */
-        dev(uart)->PSEL.RTS = UART_PIN_RTS;
-        dev(uart)->PSEL.CTS = UART_PIN_CTS;
-        REG_CONFIG |= UART_CONFIG_HWFC_Msk; /* enable HW flow control */
-    } else {
```
Comment (aabadie) [inline cpu/nrf5x_common/periph/uart.c]: This else is gone, same for non nrf52840 block

File: cpu/sam0_common/Makefile.include [hunk 11]
```diff
@@ -28,7 +28,7 @@ CFLAGS += -DDONT_USE_PREDEFINED_PERIPHERALS_HANDLERS
 # For Cortex-M cpu we use the common cortexm.ld linker script
 LINKER_SCRIPT ?= cortexm.ld
 
-# define sam0 specific pseudomodules
-PSEUDOMODULES += sam0_periph_uart_hw_fc
+# include sam0 common periph drivers
+USEMODULE += sam0_common_periph
```
Comment (aabadie) [inline cpu/sam0_common/Makefile.include]: unrelated

Comment (dylad) [reply cpu/sam0_common/Makefile.include]: I screw up during the last rebase sorry...
Fixed !

File: cpu/nrf5x_common/periph/uart.c (hunk 10 above)
Comment (dylad) [reply cpu/nrf5x_common/periph/uart.c]: Should be good.

File: cpu/nrf5x_common/periph/uart.c (hunk 9 above)
Comment (dylad) [reply cpu/nrf5x_common/periph/uart.c]: fixed.

File: cpu/nrf5x_common/periph/uart.c (hunk 8 above)
Comment (dylad) [reply cpu/nrf5x_common/periph/uart.c]: Fixed.

File: boards/cc2650stk/include/periph_conf.h (hunk 7 above)
Comment (dylad) [reply boards/cc2650stk/include/periph_conf.h]: Removed.

File: boards/cc2650stk/include/periph_conf.h (hunk 6 above)
Comment (dylad) [reply boards/cc2650stk/include/periph_conf.h]: removed.

File: boards/particle-argon/Makefile.include [hunk 12]
```diff
@@ -1,3 +1,4 @@
 USEMODULE += boards_common_particle_mesh
+USEMODULE += periph_uart_hw_fc
```
Comment (aabadie) [inline boards/particle-argon/Makefile.include]: It would be better to add this in Makefile.dep

File: boards/particle-boron/Makefile.include [hunk 13]
```diff
@@ -1,3 +1,4 @@
 USEMODULE += boards_common_particle_mesh
+USEMODULE += periph_uart_hw_fc
```
Comment (aabadie) [inline boards/particle-boron/Makefile.include]: same here

File: boards/saml21-xpro/Makefile.include [hunk 14]
```diff
@@ -1,3 +1,5 @@
 CFLAGS += -D__SAML21J18A__
 
+USEMODULE += periph_uart_hw_fc
```
Comment (aabadie) [inline boards/saml21-xpro/Makefile.include]: Please add this line in boards/saml21-xpro/Makefile.dep

File: boards/stm32f723e-disco/Makefile.dep [hunk 15]
```diff
@@ -1,4 +1,4 @@
-USEMODULE += stm32_periph_uart_hw_fc
+USEMODULE += periph_uart_hw_fc
```
Comment (aabadie) [inline boards/stm32f723e-disco/Makefile.dep]: This is OK :)

File: boards/saml21-xpro/Makefile.include (hunk 14 above)
Comment (dylad) [reply boards/saml21-xpro/Makefile.include]: I'll remove it instead. There is no need to enable this module by default on SAML21-XPRO right ?
I dunno why this line is here...

File: boards/particle-boron/Makefile.include (hunk 13 above)
Comment (dylad) [reply boards/particle-boron/Makefile.include]: done.

File: boards/particle-argon/Makefile.include (hunk 12 above)
Comment (dylad) [reply boards/particle-argon/Makefile.include]: done.

**Comment by aabadie** [comment]
please rebase

**Comment by dylad** [comment]
> please rebase

Done.

File: boards/common/remote/include/periph_common.h [hunk 16]
```diff
@@ -0,0 +1,116 @@
+/*
```
Comment (aabadie) [inline boards/common/remote/include/periph_common.h]: This file was removed recently. You had a rebase issue with this one. The `#ifdef MODULE_PERIPH_UART_HW_FC` should go here:
https://github.com/RIOT-OS/RIOT/blob/82b96ff831dff1ad1849eeecdf92b88a0c39fbd3/boards/common/remote/include/cfg_uart_default.h#L35-L51

File: cpu/nrf52/include/periph_cpu.h [hunk 17]
```diff
@@ -167,10 +167,12 @@ typedef struct {
...
-    gpio_t rx_pin;          /**< RX pin */
-    gpio_t tx_pin;          /**< TX pin */
-    gpio_t rts_pin;         /**< RTS pin - set to GPIO_UNDEF when not using HW flow control */
-    gpio_t cts_pin;         /**< CTS pin - set to GPIO_UNDEF when not using HW flow control */
+    uint8_t rx_pin;         /**< RX pin */
+    uint8_t tx_pin;         /**< TX pin */
+#ifdef MODULE_PERIPH_UART_HW_FC
+    uint8_t rts_pin;        /**< RTS pin */
+    uint8_t cts_pin;        /**< CTS pin */
+#endif
```
Comment (aabadie) [inline cpu/nrf52/include/periph_cpu.h]:
```suggestion
    gpio_t rx_pin;         /**< RX pin */
    gpio_t tx_pin;         /**< TX pin */
#ifdef MODULE_PERIPH_UART_HW_FC
    gpio_t rts_pin;        /**< RTS pin */
    gpio_t cts_pin;        /**< CTS pin */
#endif
```

**Comment by aabadie** [comment]
There are issues with Murdock, seems your rebase had problems. We should merge this one asap, otherwise it's difficult to keep the PR clean.

File: boards/particle-argon/Makefile.include (hunk 12 above)
Comment (aabadie) [reply boards/particle-argon/Makefile.include]: It seems that this comment is still unaddressed. Those 2 USEMODULE should be in Makefile.dep.

File: boards/particle-boron/Makefile.include (hunk 13 above)
Comment (aabadie) [reply boards/particle-boron/Makefile.include]: same here ;)

File: boards/saml21-xpro/Makefile.include (hunk 14 above)
Comment (aabadie) [reply boards/saml21-xpro/Makefile.include]: Not sure if this line should still be there ?

File: boards/particle-argon/Makefile.include (hunk 12 above)
Comment (dylad) [reply boards/particle-argon/Makefile.include]: Must be due to the rebase... I knew I shouldn't rebase at night...

File: boards/saml21-xpro/Makefile.include (hunk 14 above)
Comment (aabadie) [reply boards/saml21-xpro/Makefile.include]: This line is my only remaining concern: can we move it to the board `Makefile.dep` or just remove it ?

Comment (dylad) [reply boards/saml21-xpro/Makefile.include]: This one is still here ?! Damn, I'm sorry I'll remove it

**Comment by aabadie** [comment]
Thanks for updating shortly. Changes are good now, so please squash. I'll give a try to this PR later today.

File: boards/nrf52840-mdk/include/periph_conf.h [hunk 18]
```diff
@@ -40,8 +40,10 @@ static const uart_conf_t uart_config[] = {
         .dev        = NRF_UARTE0,
         .rx_pin     = GPIO_PIN(0,19),
         .tx_pin     = GPIO_PIN(0,20),
-        .rts_pin    = GPIO_UNDEF,
-        .cts_pin    = GPIO_UNDEF,
+#ifdef MODULE_PERIPH_UART_HW_FC
+        .rts_pin    = (uint8_t)GPIO_UNDEF,
+        .cts_pin    = (uint8_t)GPIO_UNDEF,
```
Comment (aabadie) [inline boards/nrf52840-mdk/include/periph_conf.h]: Ah, one last thing:

```suggestion
        .rts_pin    = GPIO_UNDEF,
        .cts_pin    = GPIO_UNDEF,
```

Comment (dylad) [reply boards/nrf52840-mdk/include/periph_conf.h]: Fixed, 
squashed directly.

**Comment by aabadie** [comment]
Ah, there are trailing whitespaces reported by Travis. You can squash directly.

**Comment by dylad** [comment]
Force-pushed again to fix a trailing whitespace issue reported by travis.

**Comment by aabadie** [comment]
And go!

**Comment by dylad** [comment]
Thanks @aabadie !
